File: app/src/main/java/education/mahmoud/quranyapp/feature/show_sura_ayas/PageAdapter.java
package education.mahmoud.quranyapp.feature.show_sura_ayas;

import android.graphics.Typeface;
import android.os.Build;
import android.os.CountDownTimer;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.text.Layout;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;

import java.text.MessageFormat;
import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import education.mahmoud.quranyapp.R;
import education.mahmoud.quranyapp.Util.Data;
import education.mahmoud.quranyapp.Util.Util;
import education.mahmoud.quranyapp.data_layer.local.room.AyahItem;

public class PageAdapter extends RecyclerView.Adapter<PageAdapter.Holder> {

    private static final String TAG = "PageAdapter";

    int ayahsColor, scrollColor;
    int vis = View.INVISIBLE;
    private List<Page> list = new ArrayList<>();
    private Typeface typeface;
    private IOnClick iOnClick;
    private PageShown pageShown;
    private IBookmark iBookmark;

    public PageAdapter(Typeface typeface, int ayahsColor, int scrollColor) {
        this.typeface = typeface;
        this.ayahsColor = ayahsColor;
        this.scrollColor = scrollColor;
    }

    public void setPageShown(PageShown pageShown) {
        this.pageShown = pageShown;
    }

    public void setiBookmark(IBookmark iBookmark) {
        this.iBookmark = iBookmark;
    }

    public void setiOnClick(IOnClick iOnClick) {
        this.iOnClick = iOnClick;
    }

    public void setPageList(List<Page> newList) {
        list = new ArrayList<>(newList);
        notifyDataSetChanged();
    }

    public List<Page> getList() {
        return list;
    }

    @Override
    public Holder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        View view = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.page_item, viewGroup, false);
        return new Holder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull Holder holder, int index) {
        Page item = list.get(index);

        holder.topLinear.setVisibility(View.VISIBLE);
        holder.BottomLinear.setVisibility(View.VISIBLE);

        // set Colors
        holder.tvAyahs.setTextColor(ayahsColor);
        holder.scAyahsText.setBackgroundColor(scrollColor);

        //<editor-fold desc="Create Text">
        StringBuilder builder = new StringBuilder();
        String aya;
        String suraName = "";
        suraName = getSuraNameFromIndex(item.getAyahItems().get(0).getSurahIndex());

        String tempSuraName;
        boolean isFirst = true;
        for (AyahItem ayahItem : item.getAyahItems()) {
            aya = ayahItem.getText();
            // add sura name
            if (ayahItem.getAyahInSurahIndex() == 1) {
                tempSuraName = getSuraNameFromIndex(ayahItem.getSurahIndex());
                if (isFirst) {
                    // handle first name in page that not need a previous new line
                    builder.append(tempSuraName + "\n");
                } else {
                    builder.append("\n" + tempSuraName + "\n");
                }

                // AlFatiha(index = 1 ) has a Basmallah in first ayah.
                if (ayahItem.getSurahIndex() != 1) {
                    int pos = aya.indexOf("ٱلرَّحِيم");
                    Log.d(TAG, "onBindViewHolder: pos " + pos);
                    pos += (new String("ٱلرَّحِيم").length());
                    Log.d(TAG, "onBindViewHolder: last text after bsmallah " + aya.substring(pos));
                    // insert  البسملة
                    builder.append(aya.substring(0, pos + 1)); // +1 as substring upper bound is excluded
                    builder.append("\n");
                    // cute ayah
                    aya = aya.substring(pos+1); // +1 to start with new character after البسملة
                }
            }
            isFirst = false;
            builder.append(MessageFormat.format("{0} ﴿ {1} ﴾ ", aya, ayahItem.getAyahInSurahIndex()));

        }
        //</editor-fold>

        holder.tvAyahs.setText(Util.getSpannable(builder.toString()), TextView.BufferType.SPANNABLE);
        holder.tvAyahs.setTypeface(typeface);

        // text justifivation
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            holder.tvAyahs.setJustificationMode(Layout.JUSTIFICATION_MODE_NONE);
        }

        // top - bottom details
        holder.tvPageNumShowAyahs.setText(String.valueOf(item.getPageNum()));
        holder.tvSurahName.setText(suraName);
        holder.tvJuz.setText(String.valueOf(item.getJuz()));

        //<editor-fold desc="listeners">
        holder.imBookmark.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                iBookmark.onBookmarkClicked(item);
            }
        });

        holder.ayahsLayout.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Log.d(TAG, "onClick: frameLayout");
                flipState(holder);
            }
        });

        holder.btnNext.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                    iOnClick.onClick(holder.getAdapterPosition());
            }
        });

        holder.btnPrev.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                    iOnClick.onClick(holder.getAdapterPosition() - 2); // there will be update by one
            }
        });
        //</editor-fold>

        //<editor-fold desc="configure next/prev buttons">
        if (index == 0 ){
            holder.btnNext.setVisibility(View.VISIBLE);
            holder.btnPrev.setVisibility(View.INVISIBLE);
        }else if (index == 603 ){
            holder.btnNext.setVisibility(View.INVISIBLE);
            holder.btnPrev.setVisibility(View.VISIBLE);
        }else{
            holder.btnNext.setVisibility(View.VISIBLE);
            holder.btnPrev.setVisibility(View.VISIBLE);
        }
        //</editor-fold>


    }

    private void flipState(Holder holder) {
        vis = holder.topLinear.getVisibility();
        vis = vis == View.VISIBLE ? View.INVISIBLE : View.VISIBLE;
        holder.BottomLinear.setVisibility(vis);
        holder.topLinear.setVisibility(vis);
    }

    /**
     * @param surahIndex in quran
     * @return
     */
    private String getSuraNameFromIndex(int surahIndex) {
        return Data.SURA_NAMES[surahIndex - 1];
    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    @Override
    public void onViewAttachedToWindow(@NonNull Holder holder) {
        super.onViewAttachedToWindow(holder);


        /* //<editor-fold desc="timer to hide">
        new CountDownTimer(1000, 1000) {
            @Override
            public void onTick(long l) {

            }

            @Override
            public void onFinish() {
                vis = View.VISIBLE;
                holder.BottomLinear.setVisibility(vis);
                holder.topLinear.setVisibility(vis);
            }
        }.start();
        //</editor-fold>*/

        pageShown.onDiplayed(holder.getAdapterPosition(), holder);
    }

    public Page getPage(int pos) {
        return list.get(pos);
    }

    interface PageShown {
        void onDiplayed(int pos, Holder holder);
    }

    interface IBookmark {
        void onBookmarkClicked(Page item);
    }

    class Holder extends RecyclerView.ViewHolder {

        @BindView(R.id.tvAyahs)
        TextView tvAyahs;
        @BindView(R.id.sc_ayahs_text)
        ScrollView scAyahsText;
        @BindView(R.id.tvSurahName)
        TextView tvSurahName;
        @BindView(R.id.tvJuz)
        TextView tvJuz;
        @BindView(R.id.imBookmark)
        ImageView imBookmark;
        @BindView(R.id.topLinear)
        LinearLayout topLinear;
        @BindView(R.id.btnNext)
        ImageButton btnNext;
        @BindView(R.id.tvPageNumShowAyahs)
        TextView tvPageNumShowAyahs;
        @BindView(R.id.btnPrev)
        ImageButton btnPrev;
        @BindView(R.id.BottomLinear)
        LinearLayout BottomLinear;
        @BindView(R.id.ayahsLayout)
        FrameLayout ayahsLayout;

        public Holder(@NonNull View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }

}

File: app/src/main/java/education/mahmoud/quranyapp/feature/show_sura_ayas/PageAdapter.kt
package education.mahmoud.quranyapp.feature.show_sura_ayas

import android.os.Build
import android.os.CountDownTimer
import android.text.Layout
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.recyclerview.widget.RecyclerView
import butterknife.BindView
import butterknife.ButterKnife
import education.mahmoud.quranyapp.R
import education.mahmoud.quranyapp.utils.Data
import education.mahmoud.quranyapp.utils.Util
import java.text.MessageFormat

class PageAdapter(var ayahsColor: Int, var scrollColor: Int) : RecyclerView.Adapter<PageAdapter.Holder>() {
    var vis = View.INVISIBLE

    var list: ArrayList<Page> = arrayListOf()

    private var iOnClick: IOnClick? = null
    private var pageShown: PageShown? = null
    private var iBookmark: IBookmark? = null
    fun setPageShown(pageShown: PageShown?) {
        this.pageShown = pageShown
    }

    fun setiBookmark(iBookmark: IBookmark?) {
        this.iBookmark = iBookmark
    }

    fun setiOnClick(iOnClick: IOnClick?) {
        this.iOnClick = iOnClick
    }

    fun setPageList(newList: List<Page>) {
        list.clear()
        list.addAll(newList)
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): Holder {
        val view = LayoutInflater.from(viewGroup.context).inflate(R.layout.page_item, viewGroup, false)
        return Holder(view)
    }

    override fun onBindViewHolder(holder: Holder, index: Int) {
        val item = list[index]
        holder.topLinear?.visibility = View.VISIBLE
        holder.BottomLinear?.visibility = View.VISIBLE
        // set Colors
        holder.tvAyahs?.setTextColor(ayahsColor)
        holder.scAyahsText?.setBackgroundColor(scrollColor)
        // <editor-fold desc="Create Text">
        val builder = StringBuilder()
        var aya: String
        var suraName = ""
        suraName = getSuraNameFromIndex(item.ayahItems[0].surahIndex)
        var tempSuraName: String
        var isFirst = true
        for (ayahItem in item.ayahItems) {
            aya = ayahItem.text
            // add sura name
            if (ayahItem.ayahInSurahIndex == 1) {
                tempSuraName = getSuraNameFromIndex(ayahItem.surahIndex)
                if (isFirst) { // handle first name in page that not need a previous new line
                    builder.append(tempSuraName + "\n")
                } else {
                    builder.append("\n" + tempSuraName + "\n")
                }
                // AlFatiha(index = 1 ) has a Basmallah in first ayah.
                if (ayahItem.surahIndex != 1) {
                    var pos = aya.indexOf("ٱلرَّحِيم")
                    Log.d(TAG, "onBindViewHolder: pos $pos")
                    pos += "ٱلرَّحِيم".length
                    Log.d(TAG, "onBindViewHolder: last text after bsmallah " + aya.substring(pos))
                    // insert  البسملة
                    builder.append(aya.substring(0, pos + 1)) // +1 as substring upper bound is excluded
                    builder.append("\n")
                    // cute ayah
                    aya = aya.substring(pos + 1) // +1 to start with new character after البسملة
                }
            }
            isFirst = false
            builder.append(MessageFormat.format("{0} ﴿ {1} ﴾ ", aya, ayahItem.ayahInSurahIndex))
        }
        // </editor-fold>
        holder.tvAyahs?.setText(Util.getSpannable(builder.toString()), TextView.BufferType.SPANNABLE)
        // text justifivation
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            holder.tvAyahs?.justificationMode = Layout.JUSTIFICATION_MODE_NONE
        }
        // top - bottom details
        holder.tvPageNumShowAyahs?.text = item.pageNum.toString()
        holder.tvSurahName?.text = suraName
        holder.tvJuz?.text = item.juz.toString()
        // <editor-fold desc="listeners">
        holder.imBookmark?.setOnClickListener { iBookmark?.onBookmarkClicked(item) }
        /*holder.ayahsLayout.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Log.d(TAG, "onClick: frameLayout");
                flipState(holder);
            }
        });
        */holder.btnNext?.setOnClickListener { iOnClick?.onClick(holder.adapterPosition) }
        holder.btnPrev?.setOnClickListener {
            iOnClick?.onClick(holder.adapterPosition - 2) // there will be update by one
        }
        // </editor-fold>
        // <editor-fold desc="configure next/prev buttons">
        if (index == 0) {
            holder.btnNext?.visibility = View.VISIBLE
            holder.btnPrev?.visibility = View.INVISIBLE
        } else if (index == 603) {
            holder.btnNext?.visibility = View.INVISIBLE
            holder.btnPrev?.visibility = View.VISIBLE
        } else {
            holder.btnNext?.visibility = View.VISIBLE
            holder.btnPrev?.visibility = View.VISIBLE
        }
        // </editor-fold>

        holder.tvAyahs?.setOnClickListener { Log.d(TAG, "onClick: ") }
    }

    private fun flipState(holder: Holder) {
        /*  vis = holder.topLinear?.visibility
          vis = if (vis == View.VISIBLE) View.INVISIBLE else View.VISIBLE
          holder.BottomLinear?.visibility = vis
          holder.topLinear?.visibility = vis*/
    }

    /**
     * @param surahIndex in quran
     * @return
     */
    private fun getSuraNameFromIndex(surahIndex: Int): String {
        return Data.SURA_NAMES[surahIndex - 1]
    }

    override fun getItemCount(): Int {
        return list.size
    }

    override fun onViewAttachedToWindow(holder: Holder) {
        super.onViewAttachedToWindow(holder)
        // <editor-fold desc="timer to hide">
        object : CountDownTimer(50, 100) {
            override fun onTick(l: Long) {}
            override fun onFinish() {
                vis = View.VISIBLE
                holder.BottomLinear?.visibility = vis
                holder.topLinear?.visibility = vis
            }
        }.start()
        // </editor-fold>
        pageShown?.onDiplayed(holder.adapterPosition, holder)
    }

    fun getPage(pos: Int): Page {
        return list[pos]
    }

    interface PageShown {
        fun onDiplayed(pos: Int, holder: Holder)
    }

    interface IBookmark {
        fun onBookmarkClicked(item: Page)
    }

    inner class Holder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        @JvmField
        @BindView(R.id.tvAyahs)
        var tvAyahs: TextView? = null
        @JvmField
        @BindView(R.id.sc_ayahs_text)
        var scAyahsText: ScrollView? = null
        @JvmField
        @BindView(R.id.tvSurahName)
        var tvSurahName: TextView? = null
        @JvmField
        @BindView(R.id.tvJuz)
        var tvJuz: TextView? = null
        @JvmField
        @BindView(R.id.imBookmark)
        var imBookmark: ImageView? = null
        @JvmField
        @BindView(R.id.topLinear)
        var topLinear: LinearLayout? = null
        @JvmField
        @BindView(R.id.btnNext)
        var btnNext: ImageButton? = null
        @JvmField
        @BindView(R.id.tvPageNumShowAyahs)
        var tvPageNumShowAyahs: TextView? = null
        @JvmField
        @BindView(R.id.btnPrev)
        var btnPrev: ImageButton? = null
        @JvmField
        @BindView(R.id.BottomLinear)
        var BottomLinear: LinearLayout? = null
        @JvmField
        @BindView(R.id.ayahsLayout)
        var ayahsLayout: FrameLayout? = null

        init {
            ButterKnife.bind(this, itemView)
        }
    }

    companion object {
        private const val TAG = "PageAdapter"
    }
}